org.eclipse.jdt.launching/launching/org/eclipse/jdt/launching/JavaRuntime.java

```diff
@@ -3460,6 +3465,120 @@ public static VMStandin createVMFromDefinitionFile(File eeFile, String name, Str
 		throw new CoreException(status);
 	}
 
+	/**
+	 * Returns the names of all packages publicly exported by the given {@link IVMInstall vmInstall}, optionally restricted to the given release
+	 * version.
+	 * <p>
+	 * The returned set contains the distinct names sorted in alphabetical order.<br>
+	 * The specified release is only considered for {@link #isModularJava(IVMInstall) modular} VM-Installs, for non-modular installs it is ignored,
+	 * and then must correspond to Java version 9 or higher.
+	 * </p>
+	 *
+	 * @param vm
+	 *            the vm-install to query
+	 * @param release
+	 *            the release to restrict the packages to, ignored for non-modular VMs. Supported values are the constants defined in
+	 *            {@link JavaCore#VERSION_9 JavaCore.VERSION_X}, e.g. {@code "9"}, {@code "10"}, {@code "11"} or so on. If {@code null}, the
+	 *            VMInstall's release is used.
+	 * @return the distinct and alphabetically sorted immutable set of names of all packages in the VM
+	 * @throws CoreException
+	 *             if an error occurs or a modular VM is queried for packages of Java-8 or lower
+	 * @since 3.22
+	 */
+	public static Set<String> getProvidedVMPackages(IVMInstall vm, String release) throws CoreException {
+		boolean isModular = JavaRuntime.isModularJava(vm);
+		release = normalizeRelease(release, isModular);
+		// If release is empty we assume no release info available
+		if (vm instanceof AbstractVMInstall vmInstall) {
+			try {
+				release = String.valueOf(release); // ConcurrentHashMap does not support null keys -> use "null" instead of null
+				Set<String> packages = vmInstall.systemPackages.computeIfAbsent(release, r -> {
+					try {
+						return querySystemPackages(vmInstall, "null".equals(r) ? null : r, isModular); //$NON-NLS-1$
+					} catch (CoreException e) {
+						throw new IllegalArgumentException(e);
+					}
+				});
+				return packages;
+			} catch (IllegalArgumentException e) {
+				if (e.getCause() instanceof CoreException coreException) {
+					throw coreException; // throw passed through exception
+				}
+				throw e;
+			}
+		}
+		return querySystemPackages(vm, release, isModular);
+	}
+
+	private static String normalizeRelease(String release, boolean isModularVM) throws CoreException {
+		if (release != null) {
```

Comment:
Could we please invert condition to avoid that entire method body is inside a single "if"?